File: agenkit-rust/examples/react_pattern.rs
//! ReAct Pattern Example
//!
//! Demonstrates the ReAct (Reasoning + Acting) pattern where agents reason
//! about actions and execute tools in an iterative loop.

use agenkit::core::{Agent, AgentError, Message, Tool, ToolResult};
use agenkit::patterns::{ReActAgent, ReActConfig};
use async_trait::async_trait;
use std::collections::HashMap;
use std::sync::Arc;

/// Simple calculator tool that performs basic arithmetic
struct CalculatorTool;

#[async_trait]
impl Tool for CalculatorTool {
    fn name(&self) -> &str {
        "calculator"
    }

    fn description(&self) -> &str {
        "Performs basic arithmetic calculations. Input should be an expression like '2+2' or '15% of 240'"
    }

    async fn execute(
        &self,
        params: HashMap<String, serde_json::Value>,
    ) -> Result<ToolResult, AgentError> {
        let input = params
            .get("input")
            .and_then(|v| v.as_str())
            .unwrap_or("");

        println!("   🧮 Calculator executing: {}", input);

        // Simple evaluation (in real code, use a proper expression evaluator)
        let result = if input.contains("+") {
            let parts: Vec<&str> = input.split('+').collect();
            if parts.len() == 2 {
                let a: f64 = parts[0].trim().parse().unwrap_or(0.0);
                let b: f64 = parts[1].trim().parse().unwrap_or(0.0);
                (a + b).to_string()
            } else {
                "Error: Invalid expression".to_string()
            }
        } else if input.contains("% of") || input.contains("percent of") {
            // Parse "X% of Y" or "X percent of Y"
            let cleaned = input.replace("percent", "%");
            let parts: Vec<&str> = cleaned.split("% of").collect();
            if parts.len() == 2 {
                let percent: f64 = parts[0].trim().parse().unwrap_or(0.0);
                let value: f64 = parts[1].trim().parse().unwrap_or(0.0);
                (value * percent / 100.0).to_string()
            } else {
                "Error: Invalid percentage expression".to_string()
            }
        } else if input.contains("*") {
            let parts: Vec<&str> = input.split('*').collect();
            if parts.len() == 2 {
                let a: f64 = parts[0].trim().parse().unwrap_or(0.0);
                let b: f64 = parts[1].trim().parse().unwrap_or(0.0);
                (a * b).to_string()
            } else {
                "Error: Invalid expression".to_string()
            }
        } else {
            "Error: Unsupported operation".to_string()
        };

        Ok(ToolResult {
            output: serde_json::json!(result),
            success: !result.starts_with("Error"),
            error: if result.starts_with("Error") {
                Some(result)
            } else {
                None
            },
            metadata: HashMap::new(),
        })
    }
}

/// Search tool that looks up information
struct SearchTool;

#[async_trait]
impl Tool for SearchTool {
    fn name(&self) -> &str {
        "search"
    }

    fn description(&self) -> &str {
        "Searches for information on a given topic. Input should be a search query."
    }

    async fn execute(
        &self,
        params: HashMap<String, serde_json::Value>,
    ) -> Result<ToolResult, AgentError> {
        let query = params
            .get("input")
            .and_then(|v| v.as_str())
            .unwrap_or("");

        println!("   🔍 Search executing: {}", query);

        // Simulate search results
        let result = match query.to_lowercase().as_str() {
            q if q.contains("rust") => {
                "Rust is a systems programming language focused on safety, speed, and concurrency."
            }
            q if q.contains("agenkit") => {
                "Agenkit is a cross-language AI agent framework supporting Python, Go, TypeScript, and Rust."
            }
            q if q.contains("react") || q.contains("reasoning") => {
                "ReAct combines reasoning and acting by interleaving thought processes with tool actions."
            }
            _ => "No relevant information found for this query.",
        };

        Ok(ToolResult {
            output: serde_json::json!(result),
            success: true,
            error: None,
            metadata: HashMap::new(),
        })
    }
}

/// Mock reasoning agent that simulates LLM-style responses
struct MockReasoningAgent {
    scenario: String,
}

#[async_trait]
impl Agent for MockReasoningAgent {
    fn name(&self) -> &str {
        "MockReasoning"
    }

    async fn process(&self, message: Message) -> Result<Message, AgentError> {
        let content = message.content_as_str().unwrap_or("");

        // Simulate different reasoning based on scenario
        let response = if self.scenario == "calculator" {
            if content.contains("What is 15% of 240") {
                if content.contains("Observation:") {
                    // We've seen the result, provide final answer
                    "Thought: I have the calculation result\nFinal Answer: 15% of 240 is 36"
                } else {
                    // First step - need to calculate
                    "Thought: I need to calculate 15% of 240\nAction: calculator\nAction Input: 15% of 240"
                }
            } else {
                "Thought: Unknown question\nFinal Answer: I don't know"
            }
        } else if self.scenario == "search" {
            if content.contains("What is Rust") {
                if content.contains("Observation:") {
                    // We've seen the search result
                    "Thought: I found information about Rust\nFinal Answer: Rust is a systems programming language focused on safety, speed, and concurrency."
                } else {
                    // Need to search
                    "Thought: I should search for information about Rust\nAction: search\nAction Input: Rust programming language"
                }
            } else {
                "Thought: Unknown question\nFinal Answer: I don't know"
            }
        } else if self.scenario == "multi-step" {
            if !content.contains("Observation:") {
                // First step
                "Thought: I should first search for what ReAct means\nAction: search\nAction Input: ReAct reasoning"
            } else if content.contains("ReAct combines") {
                // Second step - we have search result
                "Thought: Now I understand ReAct, let me calculate an example\nAction: calculator\nAction Input: 10 + 5"
            } else if content.contains("15") {
                // Third step - we have calculation result
                "Thought: I have both pieces of information\nFinal Answer: ReAct combines reasoning and acting. As an example, 10 + 5 = 15."
            } else {
                "Thought: Something went wrong\nFinal Answer: Unable to complete"
            }
        } else {
            "Thought: Unknown scenario\nFinal Answer: Error"
        };

        Ok(Message::with_text("assistant", response))
    }
}

#[tokio::main]
async fn main() -> Result<(), Box<dyn std::error::Error>> {
    println!("🎭 ReAct Pattern Example\n");

    // Example 1: Simple calculation with ReAct
    println!("{}", "=".repeat(60));
    println!("📋 Example 1: Simple Calculation");
    println!("{}", "=".repeat(60));

    let reasoning_agent = Arc::new(MockReasoningAgent {
        scenario: "calculator".to_string(),
    });
    let calculator = Arc::new(CalculatorTool) as Arc<dyn Tool>;

    let config = ReActConfig {
        agent: reasoning_agent,
        tools: vec![calculator],
        max_steps: 5,
        verbose: true,
        prompt_template: None,
    };

    let react_agent = ReActAgent::new(config)?;

    println!("\n➡️  Query: What is 15% of 240?");
    let message = Message::with_text("user", "What is 15% of 240?");
    let result = react_agent.process(message).await?;

    println!("\n✅ Result:");
    println!("{}", result.content_as_str().unwrap_or(""));

    // Example 2: Information search with ReAct
    println!("\n\n{}", "=".repeat(60));
    println!("📋 Example 2: Information Search");
    println!("{}", "=".repeat(60));

    let reasoning_agent2 = Arc::new(MockReasoningAgent {
        scenario: "search".to_string(),
    });
    let search = Arc::new(SearchTool) as Arc<dyn Tool>;

    let config2 = ReActConfig {
        agent: reasoning_agent2,
        tools: vec![search],
        max_steps: 5,
        verbose: true,
        prompt_template: None,
    };

    let react_agent2 = ReActAgent::new(config2)?;

    println!("\n➡️  Query: What is Rust?");
    let message2 = Message::with_text("user", "What is Rust?");
    let result2 = react_agent2.process(message2).await?;

    println!("\n✅ Result:");
    println!("{}", result2.content_as_str().unwrap_or(""));

    // Example 3: Multi-step reasoning with multiple tools
    println!("\n\n{}", "=".repeat(60));
    println!("📋 Example 3: Multi-Step Reasoning");
    println!("{}", "=".repeat(60));

    let reasoning_agent3 = Arc::new(MockReasoningAgent {
        scenario: "multi-step".to_string(),
    });
    let calculator3 = Arc::new(CalculatorTool) as Arc<dyn Tool>;
    let search3 = Arc::new(SearchTool) as Arc<dyn Tool>;

    let config3 = ReActConfig {
        agent: reasoning_agent3,
        tools: vec![calculator3, search3],
        max_steps: 10,
        verbose: true,
        prompt_template: None,
    };

    let react_agent3 = ReActAgent::new(config3)?;

    println!("\n➡️  Query: Explain ReAct and show a calculation example");
    let message3 = Message::with_text("user", "Explain ReAct and show a calculation example");
    let result3 = react_agent3.process(message3).await?;

    println!("\n✅ Result:");
    println!("{}", result3.content_as_str().unwrap_or(""));

    println!("\n✨ ReAct pattern examples complete!\n");
    println!("💡 Key takeaways:");
    println!("   - ReAct combines reasoning (thought) with acting (tool use)");
    println!("   - Agent iterates: Thought → Action → Observation → repeat");
    println!("   - Tools extend agent capabilities beyond pure language");
    println!("   - Transparent reasoning process visible in verbose mode");

    Ok(())
}

File: examples/patterns/react-pattern.py
"""
ReAct Agent Example

Demonstrates how to use ReActAgent (Reasoning + Acting) to build agents
that can use tools to accomplish tasks.

The ReAct pattern enables agents to:
- Reason about which tools to use
- Execute tools with appropriate parameters
- Observe results and adjust strategy
- Provide final answers based on tool outputs

This example uses mock implementations for demonstration. In production,
replace with real LLM agents and tools.
"""

import asyncio
from datetime import datetime

from agenkit import Message
from agenkit.patterns import ReActAgent

# ============================================================================
# Mock Tools for Demonstration
# ============================================================================


class Calculator:
    """
    Mock calculator tool for mathematical operations.

    In production, this could be replaced with a real calculator,
    Python eval (with proper sandboxing), or a math API.
    """

    name = "calculator"
    description = (
        "Performs mathematical calculations. Input should be a mathematical expression as a string."
    )

    async def execute(self, input: str) -> str:
        """Execute a calculation using safe AST evaluation."""
        import ast
        import operator

        # Safe operations mapping
        safe_ops = {
            ast.Add: operator.add,
            ast.Sub: operator.sub,
            ast.Mult: operator.mul,
            ast.Div: operator.truediv,
            ast.Pow: operator.pow,
            ast.USub: operator.neg,
        }

        def safe_eval(node):
            """Safely evaluate a math expression AST node."""
            if isinstance(node, ast.Constant):
                return node.value
            elif isinstance(node, ast.BinOp):
                left = safe_eval(node.left)
                right = safe_eval(node.right)
                op = safe_ops.get(type(node.op))
                if op is None:
                    raise ValueError(f"Unsupported operation: {type(node.op).__name__}")
                return op(left, right)
            elif isinstance(node, ast.UnaryOp):
                operand = safe_eval(node.operand)
                op = safe_ops.get(type(node.op))
                if op is None:
                    raise ValueError(f"Unsupported operation: {type(node.op).__name__}")
                return op(operand)
            else:
                raise ValueError(f"Unsupported expression: {type(node).__name__}")

        try:
            expr = str(input).strip()
            tree = ast.parse(expr, mode="eval")
            result = safe_eval(tree.body)
            return str(result)
        except (SyntaxError, ValueError, ZeroDivisionError, TypeError) as e:
            raise ValueError(f"Invalid expression: {e}")


class WebSearch:
    """
    Mock web search tool.

    In production, replace with real search API (Google, Bing, DuckDuckGo, etc.)
    """

    name = "search"
    description = "Searches the web for information. Input should be a search query string."

    def __init__(self):
        # Mock search results database
        self.mock_results = {
            "python": "Python is a high-level programming language created by Guido van Rossum in 1991.",
            "weather": "The weather is currently sunny with a temperature of 72°F.",
            "eiffel tower": "The Eiffel Tower is a wrought-iron lattice tower in Paris, France. It was completed in 1889 and stands 330 meters tall.",
            "population of france": "The population of France is approximately 67 million people as of 2023.",
        }

    async def execute(self, input: str) -> str:
        """Perform a search."""
        query = input.lower()

        # Find matching mock result
        for key, value in self.mock_results.items():
            if key in query:
                return value

        return f"No results found for '{input}'. (This is a mock search tool)"


class DateTimeTool:
    """
    Tool for getting current date and time information.
    """

    name = "datetime"
    description = "Gets the current date and time. No input required."

    async def execute(self, **kwargs) -> str:
        """Get current datetime."""
        now = datetime.now()
        return now.strftime("%Y-%m-%d %H:%M:%S")


# ============================================================================
# Mock Agent
# ============================================================================


class MockReActAgent:
    """
    Mock Agent that simulates ReAct-style reasoning.

    In production, replace with real LLM-based agent:
    - OpenAI: from openai import AsyncOpenAI
    - Anthropic: from anthropic import AsyncAnthropic
    - LiteLLM: from litellm import acompletion
    """

    def __init__(self):
        self.call_count = 0
        self.name = "MockReActAgent"

    async def process(self, message: Message) -> Message:
        """Generate mock ReAct-style responses."""
        self.call_count += 1

        content = message.content.lower()

        # Check if this is an observation (tool result)
        if content.startswith("observation:"):
            # We have a tool result, decide on next action
            if "error" in content:
                return Message(
                    role="assistant",
                    content="Thought: The tool encountered an error\nAction: Final Answer\nAction Input: I encountered an error while processing your request.",
                )
            else:
                # Tool succeeded, provide final answer
                observation = content.replace("observation:", "").strip()
                return Message(
                    role="assistant",
                    content=f"Thought: I have the information I need\nAction: Final Answer\nAction Input: Based on the information, {observation}",
                )

        # Initial reasoning based on user query
        if "calculate" in content or "%" in content or "+" in content or "*" in content:
            # Extract math expression
            if "15% of 240" in content:
                return Message(
                    role="assistant",
                    content="Thought: I need to calculate 15% of 240\nAction: calculator\nAction Input: 240 * 0.15",
                )
            elif "what is" in content and ("+" in content or "*" in content):
                # Try to extract expression
                parts = content.split("what is")
                if len(parts) > 1:
                    expr = parts[1].strip().rstrip("?")
                    return Message(
                        role="assistant",
                        content=f"Thought: I need to calculate this expression\nAction: calculator\nAction Input: {expr}",
                    )

        elif "search" in content or "find" in content or "what is python" in content:
            # Extract search query
            if "python" in content:
                return Message(
                    role="assistant",
                    content="Thought: I should search for information about Python\nAction: search\nAction Input: python",
                )
            elif "eiffel tower" in content:
                return Message(
                    role="assistant",
                    content="Thought: I should search for information about the Eiffel Tower\nAction: search\nAction Input: eiffel tower",
                )

        elif "time" in content or "date" in content:
            return Message(
                role="assistant",
                content="Thought: I need to get the current date and time\nAction: datetime\nAction Input: {}",
            )

        # Default: provide a simple answer
        return Message(
            role="assistant",
            content="Thought: I can answer this directly\nAction: Final Answer\nAction Input: I understand your question, but I'm a demo agent with limited capabilities.",
        )


# ============================================================================
# Example Functions
# ============================================================================


async def basic_tool_usage_example():
    """Demonstrate basic tool usage with ReActAgent."""
    print("=" * 60)
    print("Example 1: Basic Tool Usage")
    print("=" * 60)

    # Setup tools
    tools = [Calculator(), WebSearch(), DateTimeTool()]

    # Create agent
    mock_agent = MockReActAgent()
    agent = ReActAgent(agent=mock_agent, tools=tools, max_steps=5)

    # Test calculation
    print("\nUser: What is 15% of 240?")
    response = await agent.process(Message(role="user", content="What is 15% of 240?"))
    print(f"Agent: {response.content}")

    # Show reasoning steps
    print(f"\nReasoning steps: {len(agent.get_steps())}")
    for step in agent.get_steps():
        print(f"  Step {step.step_number + 1}: {step.action} -> {step.observation}")


async def multiple_tools_example():
    """Demonstrate using multiple different tools."""
    print("\n" + "=" * 60)
    print("Example 2: Multiple Tools")
    print("=" * 60)

    tools = [Calculator(), WebSearch(), DateTimeTool()]

    mock_agent = MockReActAgent()
    agent = ReActAgent(agent=mock_agent, tools=tools, max_steps=5)

    queries = [
        "What is Python?",
        "What time is it?",
        "Tell me about the Eiffel Tower",
    ]

    for query in queries:
        print(f"\nUser: {query}")
        response = await agent.process(Message(role="user", content=query))
        print(f"Agent: {response.content}")


async def verbose_mode_example():
    """Demonstrate verbose mode that shows thought process."""
    print("\n" + "=" * 60)
    print("Example 3: Verbose Mode (Show Reasoning)")
    print("=" * 60)

    tools = [Calculator()]

    mock_agent = MockReActAgent()
    agent = ReActAgent(agent=mock_agent, tools=tools, max_steps=5, verbose=True)

    print("\nUser: Calculate 123 + 456")
    response = await agent.process(Message(role="user", content="Calculate 123 + 456"))
    print(f"Agent:\n{response.content}")


async def error_handling_example():
    """Demonstrate error handling when tools fail."""
    print("\n" + "=" * 60)
    print("Example 4: Error Handling")
    print("=" * 60)

    tools = [Calculator()]

    mock_agent = MockReActAgent()
    agent = ReActAgent(agent=mock_agent, tools=tools, max_steps=3)

    # This will cause the calculator to error
    print("\nUser: Calculate invalid expression")
    response = await agent.process(Message(role="user", content="Calculate 2 + + 3"))
    print(f"Agent: {response.content}")

    # Check if any errors occurred in steps
    for step in agent.get_steps():
        if "error" in step.observation.lower():
            print(f"\n✗ Step {step.step_number + 1} had an error:")
            print(f"  Action: {step.action}")
            print(f"  Error: {step.observation}")


async def custom_system_prompt_example():
    """Demonstrate using a custom system prompt."""
    print("\n" + "=" * 60)
    print("Example 5: Custom System Prompt")
    print("=" * 60)

    tools = [Calculator()]

    custom_prompt = """You are a friendly math tutor that helps students learn.

Available tools:
- calculator: Performs mathematical calculations

When solving problems:
1. Explain your reasoning clearly
2. Show the calculation steps
3. Provide the final answer

Use this format:
Thought: [your reasoning]
Action: [tool name or "Final Answer"]
Action Input: [tool input or final answer]
"""

    mock_agent = MockReActAgent()
    agent = ReActAgent(
        agent=mock_agent,
        tools=tools,
        max_steps=5,
        system_prompt=custom_prompt,
    )

    print("\nUser: Help me calculate 25 * 4")
    response = await agent.process(Message(role="user", content="Help me calculate 25 * 4"))
    print(f"Agent: {response.content}")


async def main():
    """Run all examples."""
    await basic_tool_usage_example()
    await multiple_tools_example()
    await verbose_mode_example()
    await error_handling_example()
    await custom_system_prompt_example()

    print("\n" + "=" * 60)
    print("Examples Complete!")
    print("=" * 60)
    print("\nKey Takeaways:")
    print("1. ReActAgent uses a Thought->Action->Observation loop")
    print("2. Tools are provided as a list to the agent")
    print("3. Agent automatically decides which tools to use")
    print("4. Reasoning steps can be inspected with get_steps()")
    print("5. Verbose mode shows the thought process")
    print("6. Tool errors are handled gracefully")
    print("\nNext Steps:")
    print("- Replace MockReActAgent with real LLM-based agent (OpenAI, Anthropic, etc.)")
    print("- Add real tools (APIs, databases, file systems)")
    print("- Implement custom tools for your use case")
    print("- Add retry logic for failed tool executions")
    print("- Combine with ConversationalAgent for multi-turn tool use")


if __name__ == "__main__":
    asyncio.run(main())
